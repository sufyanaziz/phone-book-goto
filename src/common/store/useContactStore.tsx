import { createContext, useState } from "react";

type TContactStore = {
  contact: { [x: string]: unknown };
  search: string;
  offset: number;
  modal: {
    name: string;
    data: { [key: string]: any };
  };
  onSetContact: (contact: { [x: string]: unknown }) => void;
  onSetSearch: (search: string) => void;
  onSetOffset: (offset: number) => void;
  closeModal: () => void;
  openModal: (modalName: string, modalData: any) => void;
};

const ContactStore = createContext<TContactStore>({
  contact: {},
  search: "",
  offset: 0,
  modal: {
    name: "",
    data: {},
  },
  onSetContact: () => {},
  onSetSearch: () => {},
  onSetOffset: () => {},
  closeModal: () => {},
  openModal: () => {},
});

type TContactProvider = {
  children?: React.ReactNode;
};

const ContactProvider = ({ children }: TContactProvider) => {
  const [contact, setContact] = useState({});
  const [search, setSearch] = useState<string>("");
  const [offset, setOffset] = useState<number>(0);
  const [modal, setModal] = useState({ name: "", data: {} });
  return (
    <ContactStore.Provider
      value={{
        contact,
        onSetContact: setContact,
        search,
        onSetSearch: (search) => {
          setSearch(search);
          setOffset(0);
        },
        offset,
        onSetOffset: setOffset,
        modal,
        openModal: (modalName, modalData) => {
          setModal({ name: modalName, data: modalData });
        },
        closeModal: () => {
          setModal({ name: "", data: {} });
        },
      }}
    >
      {children}
    </ContactStore.Provider>
  );
};

const ContactConsumer = ContactStore.Consumer;

export { ContactStore, ContactProvider, ContactConsumer };
