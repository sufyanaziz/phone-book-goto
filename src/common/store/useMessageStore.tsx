import { TMessageId } from "@common/types/message";
import { createContext, useState } from "react";

type TMessageStore = {
  showMessage: boolean;
  activeId: TMessageId;
  setShowMessage: (activeId: TMessageId, showMessage: boolean) => void;
};

const MessageStore = createContext<TMessageStore>({
  showMessage: false,
  activeId: "",
  setShowMessage: () => {},
});

type TMessageProvider = {
  children?: React.ReactNode;
};

const MessageProvider = ({ children }: TMessageProvider) => {
  const [message, setMessage] = useState<{
    showMessage: boolean;
    activeId: TMessageId;
  }>({
    showMessage: false,
    activeId: "",
  });

  return (
    <MessageStore.Provider
      value={{
        showMessage: message.showMessage,
        activeId: message.activeId,
        setShowMessage: (activeId, showMessage) => {
          setMessage({
            activeId,
            showMessage,
          });
        },
      }}
    >
      {children}
    </MessageStore.Provider>
  );
};

const ContactConsumer = MessageStore.Consumer;

export { MessageStore, MessageProvider, ContactConsumer };
