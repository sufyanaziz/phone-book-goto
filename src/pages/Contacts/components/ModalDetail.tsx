import React, { useContext, useState, useEffect } from "react";
import Modal from "@common/components/Modal";
import { ContactStore } from "@common/store/useContactStore";
import { MODAL_NAME } from "../constant/modal";
import Text from "@common/components/Text";
import { css } from "@emotion/react";
import Button from "@common/components/Button";
import { AddSquareIcon, EditIcon, MinusSquareIcon } from "@assets/icon";
import Input from "@common/components/Input";

type TModalDetail = {
  onAddNewNumber: (id: number, number: string) => void;
  onEditPhoneNumber: (
    oldData: { id: number; number: string },
    newNumber: string
  ) => void;
};

const ModalDetail = ({ onAddNewNumber, onEditPhoneNumber }: TModalDetail) => {
  const [isAddNewPhone, setIsAddNewPhone] = useState(false);
  const [newNumber, setNewNumber] = useState("");
  const [phones, setPhones] = useState([""]);
  const [editNumber, setEditNumber] = useState({
    isEdit: false,
    idx: 0,
  });
  const { modal, closeModal, openModal } = useContext(ContactStore);

  const modalDetailContactStyle = () => {
    return css({
      padding: "12px 0px",
      ".section-name": {
        display: "flex",
        gap: "4px",
        width: "100%",
        ".first-name, .last-name": {
          flex: "1",
          ".header-title": {
            marginBottom: "4px",
          },
        },
        marginBottom: "12px",
      },
      ".section-phone-number": {
        ".header-container": {
          display: "flex",
          alignItems: "center",
          marginBottom: "4px",
          ".header-title": {
            marginRight: "4px",
          },
        },
        ".phone": {
          marginBottom: "4px",
          "&:last-child": {
            marginBottom: 0,
          },
        },
      },
    });
  };

  useEffect(() => {
    if (modal.name === MODAL_NAME.modalDetailContact) {
      const resPhones = modal.data.phones?.map(
        (phone: { number: string }) => phone.number
      ) || [""];
      setPhones(resPhones);
    }
  }, [modal.name]);

  const onCloseModal = () => {
    closeModal();
    setIsAddNewPhone(false);
    setNewNumber("");
    setPhones([""]);
    setEditNumber({ isEdit: false, idx: 0 });
  };

  return (
    <Modal
      isVisible={modal.name === MODAL_NAME.modalDetailContact}
      header={
        <div css={{ display: "flex", alignItems: "center" }}>
          <Text
            className="header-title"
            text="Detail Contact"
            type="medium"
            size="medium"
          />
          <div css={{ marginLeft: "8px" }}>
            <Button
              label="Edit"
              type="warning"
              onClick={() => openModal(MODAL_NAME.modalEditContact, modal.data)}
            />
          </div>
        </div>
      }
      onClose={onCloseModal}
      content={
        <div css={modalDetailContactStyle}>
          <div className="section-name">
            <div className="first-name">
              <Text className="header-title" text="First Name" type="bold" />
              <Text text={modal.data.first_name || "-"} />
            </div>
            <div className="last-name">
              <Text className="header-title" text="Last Name" type="bold" />
              <Text text={modal.data.last_name || "-"} />
            </div>
          </div>
          <div className="section-phone-number">
            <div className="header-container">
              <Text className="header-title" text="Phones" type="bold" />
              {!editNumber.isEdit ? (
                <div onClick={() => setIsAddNewPhone(!isAddNewPhone)}>
                  {isAddNewPhone ? (
                    <MinusSquareIcon height="12px" width="12px" />
                  ) : (
                    <AddSquareIcon height="12px" width="12px" />
                  )}
                </div>
              ) : null}
            </div>
            {phones?.map((phone, idx) => (
              <div
                className="phone"
                css={{ display: "flex", alignItems: "center" }}
                key={phone}
              >
                <Text key={phone} text={phone} />
                {isAddNewPhone ||
                (editNumber.idx !== idx && editNumber.isEdit) ? null : (
                  <div
                    css={{ marginLeft: "4px" }}
                    onClick={() => {
                      setEditNumber({ isEdit: !editNumber.isEdit, idx });
                      if (editNumber.isEdit) {
                        setNewNumber("");
                      } else {
                        setNewNumber(phone);
                      }
                    }}
                  >
                    <EditIcon height="12px" width="12px" />
                  </div>
                )}
              </div>
            )) || "-"}
            {isAddNewPhone || editNumber.isEdit ? (
              <div css={{ display: "flex" }}>
                <Input
                  value={newNumber}
                  acceptValue="number-only"
                  onChange={setNewNumber}
                />
                <div css={{ marginLeft: "4px" }}>
                  <Button
                    label={isAddNewPhone ? "Add" : "Edit"}
                    disabled={!newNumber}
                    onClick={() => {
                      if (isAddNewPhone) {
                        setPhones((prev) => [...prev, newNumber]);
                        setIsAddNewPhone(false);
                        setNewNumber("");
                        onAddNewNumber(modal.data.id, newNumber);
                      } else {
                        const tempPhones = [...phones];
                        tempPhones[editNumber.idx] = newNumber;
                        setPhones(tempPhones);
                        const oldNumber = phones[editNumber.idx];
                        onEditPhoneNumber(
                          { id: modal.data.id, number: oldNumber },
                          newNumber
                        );
                        setNewNumber("");
                        setEditNumber({ isEdit: false, idx: 0 });
                      }
                    }}
                  />
                </div>
              </div>
            ) : null}
          </div>
        </div>
      }
    />
  );
};

export default ModalDetail;
