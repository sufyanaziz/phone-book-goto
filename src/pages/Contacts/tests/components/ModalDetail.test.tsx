import React from "react";
import renderer from "react-test-renderer";
import ModalDetail from "@pages/Contacts/components/ModalDetail";
import { ContactStore } from "@common/store/useContactStore";
import { mockStoreData } from "../mock/mockStore";

describe("Component ModalDetail", () => {
  it("should render ModalDetail component correctly", () => {
    const component = renderer.create(
      <ContactStore.Provider
        value={{
          ...mockStoreData,
          modal: {
            name: "modal-detail-contact",
            data: {
              first_name: "test 123",
              last_name: "last 1234",
              phones: [{ number: "6785421" }],
            },
          },
        }}
      >
        <ModalDetail />
      </ContactStore.Provider>
    );
    expect(component).toMatchSnapshot();
  });
});
