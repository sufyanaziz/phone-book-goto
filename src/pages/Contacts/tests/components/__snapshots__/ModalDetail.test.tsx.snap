// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component ModalDetail should render ModalDetail component correctly 1`] = `
<div
  css={[Function]}
>
  <div
    className="modal-mask"
    onClick={[Function]}
  />
  <div
    className="modal-content"
  >
    <div
      className="header"
    >
      <div
        css={
          {
            "alignItems": "center",
            "display": "flex",
          }
        }
      >
        <div
          className="header-title"
          css={[Function]}
        >
          Detail Contact
        </div>
        <div
          css={
            {
              "marginLeft": "8px",
            }
          }
        >
          <button
            className="warning"
            css={[Function]}
            onClick={[Function]}
          >
            Edit
          </button>
        </div>
      </div>
      <div
        className="close-icon"
        onClick={[Function]}
      >
        <div
          className=""
          css={[Function]}
        >
          X
        </div>
      </div>
    </div>
    <div
      className="content"
    >
      <div
        css={[Function]}
      >
        <div
          className="section-name"
        >
          <div
            className="first-name"
          >
            <div
              className="header-title"
              css={[Function]}
            >
              First Name
            </div>
            <div
              className=""
              css={[Function]}
            >
              test 123
            </div>
          </div>
          <div
            className="last-name"
          >
            <div
              className="header-title"
              css={[Function]}
            >
              Last Name
            </div>
            <div
              className=""
              css={[Function]}
            >
              last 1234
            </div>
          </div>
        </div>
        <div
          className="section-phone-number"
        >
          <div
            className="header-title"
            css={[Function]}
          >
            Phones
          </div>
          <div
            className="phone"
            css={[Function]}
          >
            6785421
          </div>
        </div>
      </div>
    </div>
    <div
      className="footer"
    />
  </div>
</div>
`;
